doc/modbus_reply_callback.txt

```diff
@@ -0,0 +1,160 @@
+Each callback is given the user-context, the slave-address (useful in RTU),
+the access type of function, the request start-address and the number of
+values (not bytes) to be treated. *read* and *write* additionally get the byte-buffer
+from the request-buffer at the offset where the payload begins.
+
+*verify* is used by *modbus_reply_callback()* to verify whether a request
+(read or write) is valid, the user has to return in his implementation either
+0 for success or EMBXILADD in case an access is out-of-bounds. *modbus_reply_callback*
+does modbus-specific checks on its own. *verify* basically serves for
+validating whether the address-range is valid for the requested function.
+
+*read* is used for any read access. Based on the function-argument the user has to do
+the de-serialization from the payload format to useful data. Coils/bits are stuffed into bytes,
+registers are combined over 2 bytes in big-endian byte-order (MSB first).
+
+*write* is used for any write access. Based on the function-argument the user has to do
+the serialization into the payload buffer. The write-callback has to return the number of bytes
+added to the buffer. For example having added 9 bits two bytes have been filled and thus
+the return-value is two.
+
+An optional parameter _user_ctx_ can be suppied which will be passed to
+the callback function, typically for context purpose.
+
+This function is designed for Modbus TCP server and RTU slaves.
+
+RETURN VALUE
+------------
+The function shall return the length of the response sent if
+successful. Otherwise it shall return -1 and set errno.
+
+ERRORS
+------
+*EMBMDATA*::
+Sending has failed
+
+See also the errors returned by the syscall used to send the response (eg. send
+or write).
+
+EXAMPLE
+-------
+
+The 'modbus_reply' function (based on modbus_mapping_t) is using the 'modbus_reply_callback'
+implementation for all read and write ModBus-function and is there a good example. It can be found
+in modbus.c .
+
+[source,c]
+-------------------
+
+static int my_verify(void *user_ctx, int slave, int function,
+					 uint16_t address, int nb)
+{
+	int is_input = 0;
+
+	switch (function) {
+	case MODBUS_FC_READ_DISCRETE_INPUTS:
+		is_input = 1; /* fall-through */
+	case MODBUS_FC_READ_COILS:
+	case MODBUS_FC_WRITE_SINGLE_COIL:
+	case MODBUS_FC_WRITE_MULTIPLE_COILS:
+		if (address > 207)
+			return EMBXILVAL;
+		break;
+
+	case MODBUS_FC_READ_INPUT_REGISTERS:
+		is_input = 1; /* fall-through */
+	case MODBUS_FC_READ_HOLDING_REGISTERS:
+	case MODBUS_FC_WRITE_SINGLE_REGISTER:
+	case MODBUS_FC_WRITE_MULTIPLE_REGISTERS:
+	case MODBUS_FC_MASK_WRITE_REGISTER:
+	case MODBUS_FC_WRITE_AND_READ_REGISTERS:
+		if (address > 105)
+			return EMBXILVAL;
+		break;
+	}
+
+	return 0;
+}
+
+static int my_read(void *user_ctx, int slave, int function,
+				   uint16_t address, int nb, uint8_t bytes[], int len)
+{
+	application_context *ctx = user_ctx;
+
+	/* ... */
+
```

Comment:
Agreed, the doc needs some more work. Have you looked at the implementation I put into modbus.c for the mapping-reply?

Modbus internals yes, but not many. A user needs to know how to get/put bits and registers out of/into the buffer given to the read and write callback. I chose to propose it like this because it avoids allocating  memory and additional copies inside libmodbus.